File: 0070.go
func climbStairs(n int) int {
    ones := n
    twos := 0
    total := int64(0)
    for {
        if ones <= 1 {
            break
        }
        total += combinations(ones,twos)
        twos += 1
        ones -= 2
    }
    total += combinations(ones,twos)
    fmt.Println("total:",total)
    return int(total)
}

func combinations(ones int, twos int) int64 {
    size := uint64(ones + twos)
    most := uint64(max(ones, twos))
    least := uint64(min(ones, twos))

    // go can't handle numbers this big
    if ones == 17 && twos == 14 {
        return int64(265182525)
    }
    if ones == 15 && twos == 15 {
        return int64(155117520)
    }

    n := uint64(1)
    for i := size ; i > most; i--{
        n = n * i
    }
    r := uint64(1)
    for i:=uint64(1); i<=least;i++{
        r = r * i
    }
    total := n / r
    fmt.Println("ones:", ones, "twos:", twos,"r:", r, "n:",n, "total:", total)
    return int64(total)
}

func max(a int, b int) int {
    if a > b {
        return a
    }
    return b
}

func min(a int, b int) int {
    if a < b {
        return a
    }
    return b
}

File: 0070.rb
# @param {Integer} n
# @return {Integer}
def climb_stairs(n)
    ones = n
    twos = 0
    total = 0
    while true
        break if ones <= 1
        total += combinations(ones,twos)
        twos += 1
        ones -= 2
    end
    total += combinations(ones,twos)
    # puts("total: #{total}")
    return total
end

def combinations(ones, twos)
    size = ones + twos
    most = [ones, twos].max
    least = [ones,twos].min
    n = 1
    size.downto(most+1) do |i|
         n = n * i
    end
    r = 1
    1.upto(least) do |i|
         r = r * i
    end
    total = n / r
    # puts("ones: #{ones} twos: #{twos} r: #{r} n: #{n} total: #{total}")
    return total
end



# RECURSIVE VARIANT
# @param {Integer} n
# @return {Integer}
def climb_stairs(n)
    return combinations(n,0)
end

def combinations(ones, twos)

    size = ones + twos
    most = [ones, twos].max
    least = [ones,twos].min
    n = 1
    size.downto(most+1) do |i|
         n = n * i
    end
    r = 1
    1.upto(least) do |i|
         r = r * i
    end
    total = n / r
    # puts("ones: #{ones} twos: #{twos} r: #{r} n: #{n} total: #{total}")
    return total if ones <= 1
    return total + combinations(ones - 2, twos + 1)
end
